server/base/src/main/java/org/apache/accumulo/server/conf/codec/VersionedProperties.java

```diff
@@ -0,0 +1,258 @@
+    } else {
+      this.props = new ImmutableMap.Builder<String,String>().build();
+    }
+  }
+
+  /**
+   * Get an unmodifiable map with all property key,value pairs.
+   *
+   * @return An unmodifiable view of the property key, value pairs.
+   */
+  public Map<String,String> getProperties() {
+    return props;
+  }
+
+  /**
+   * Get the current data version. The version should match the node version of the stored data. The
+   * value should be used on data writes as the expected version. If the data write fails do to an
+   * unexpected version, it signals that the node version has changed since the instance was
+   * instantiated and encoded.
+   *
+   * @return 0 for initial version, otherwise the data version when the properties were serialized.
+   */
+  public int getDataVersion() {
+    return Math.max(dataVersion, 0);
+  }
+
+  /**
+   * Calculates the version that should be stored when serialized. The serialized version, when
+   * stored, should match the version that will be assigned. This way, data reading the serialized
+   * version can compare the stored version with the node version at any time to detect if the node
+   * version has been updated.
+   * <p>
+   * The initialization of the data version to a negative value allows this value to be calculated
+   * correctly for the first serialization. On the first store, the expected version will be 0.
+   *
+   * @return the next version number that should be serialized, or 0 if this is the initial version.
+   */
+  public int getNextVersion() {
+    return Math.max(dataVersion + 1, 0);
+  }
+
+  /**
+   * The timestamp of the instance when created or last modified.
+   *
+   * @return the timestamp of the instance.
+   */
+  public Instant getTimestamp() {
+    return Instant.from(timestamp);
+  }
+
+  /**
+   * The timestamp formatted as an ISO 8601 string with format of
+   * {@code YYYY-MM-DDTHH:mm:ss.SSSSSSZ}
+   *
+   * @return a formatted timestamp string.
+   */
+  public String getTimestampISO() {
+    return tsFormatter.format(timestamp);
+  }
+
+  /**
+   * Update a single property. If a property already exists it is overwritten.
+   * <p>
+   * It is much more efficient to add multiple properties at a time rather than one by one.
+   * <p>
+   * Because instances of this class are immutable, this method creates a new copy of the
+   * properties. Other processes will continue to see original values retrieved from the data store.
+   * Other processes will receive an update when the instance is encoded and stored in the data
+   * store and then retrieved with the normal store update mechanisms.
+   *
+   * @param key
+   *          the property name.
+   * @param value
+   *          the property value.
+   * @return A new instance of this class with the property added or updated.
+   */
+  public VersionedProperties addOrUpdate(final String key, final String value) {
+    ImmutableMap<String,String> updated =
```

Comment:
You could just call the other addOrUpdate method here, making a map with a single k,v pair.

Reply:
This avoids creating an additional hashmap - it would only be one entry and not a big deal - it may turn out that this method could be refactored out as upstream changes could do the same thing and just call addOrUpdate(map) - it will depend on how many places set just one property and if this is really a convenience or not.